File: C#/Testing/CRSpeedTest/CRSpeedTest.cs
//-----------------------------------------------------------------------------
// CRSpeedTest.cs
//
// Performs LabJack operations in a loop and reports the timing statistics for
// the operations.
//
// support@labjack.com
// Nov. 21, 2013
//-----------------------------------------------------------------------------

using System;
using System.Diagnostics;
using LabJack;

namespace CRSpeedTest
{
    class CRSpeedTest
    {
        static void Main(string[] args)
        {
            CRSpeedTest spTest = new CRSpeedTest();
            spTest.performActions();
        }

        public void showErrorMessage(LJM.LJMException e)
        {
            Console.Out.WriteLine("LJMException: " + e.ToString());
            Console.Out.WriteLine(e.StackTrace);
        }

        public void performActions()
        {
            int handle = 0;
            int devType = 0;
            int conType = 0;
            int serNum = 0;
            int ipAddr = 0;
            int port = 0;
            int maxBytesPerMB = 0;
            string ipAddrStr = "";
            int i = 0;

            try
            {
                //Open first found LabJack
                LJM.OpenS("ANY", "ANY", "ANY", ref handle);
                //devType = LJM.CONSTANTS.dtANY; //Any device type
                //conType = LJM.CONSTANTS.ctANY; //Any connection type
                //LJM.Open(devType, conType, "ANY", ref handle);

                LJM.GetHandleInfo(handle, ref devType, ref conType, ref serNum, ref ipAddr, ref port, ref maxBytesPerMB);
                LJM.NumberToIP(ipAddr, ref ipAddrStr);
                Console.WriteLine("Opened a LabJack with Device type: " + devType + ", Connection type: " + conType + ",");
                Console.WriteLine("Serial number: " + serNum + ", IP address: " + ipAddrStr + ", Port: " + port + ",");
                Console.WriteLine("Max bytes per MB: " + maxBytesPerMB);


                const int numIterations = 1000; // Number of iterations to perform in the loop

                //Analog input settings
                const int numAIN = 1; // Number of analog inputs to read
                const double rangeAIN = 10.0;
                const double resolutionAIN = 1.0;

                //Digital settings
                const bool readDigital = false;
                const bool writeDigital = false;

                //Analog output settings
                const bool writeDACs = false;

                //Variables for LJM library calls
                int numFrames = 0;
                string[] aNames;
                int[] aWrites;
                int[] aNumValues;
                double[] aValues;
                int errAddr = 0;

                if(numAIN > 0)
                {
                    //Configure analog input settings
                    numFrames = Math.Max(0, numAIN*2);
                    aNames = new string[numFrames];
                    aValues = new double[numFrames];
                    for(i = 0; i < numAIN; i++)
                    {
                        aNames[i*2] = "AIN" + i + "_RANGE";
                        aValues[i*2] = rangeAIN;
                        aNames[i*2+1] = "AIN" + i + "_RESOLUTION_INDEX";
                        aValues[i*2+1] = resolutionAIN;
                    }
                    LJM.eWriteNames(handle, numFrames, aNames, aValues, ref errAddr);
                }

                //Initialize and configure eNames parameters for loop's eNames call
                numFrames = Math.Max(0, numAIN) + Convert.ToInt32(readDigital) +
                    Convert.ToInt32(writeDigital) + Convert.ToInt32(writeDACs)*2;
                aNames = new string[numFrames];
                aWrites = new int[numFrames];
                aNumValues = new int[numFrames];
                aValues = new double[numFrames]; //In this case numFrames is the size of aValue

                //Add analog input reads (AIN 0 to numAIN-1)
                for(i = 0; i < numAIN; i++)
                {
                    aNames[i] = "AIN" + i;
                    aWrites[i] = LJM.CONSTANTS.READ;
                    aNumValues[i] = 1;
                    aValues[i] = 0;
                }

                if(readDigital)
                {
                    //Add digital read
                    aNames[i] = "DIO_STATE";
                    aWrites[i] = LJM.CONSTANTS.READ;
                    aNumValues[i] = 1;
                    aValues[i] = 0;
                    i++;
                }

                if(writeDigital)
                {
                    //Add digital write
                    aNames[i] = "DIO_STATE";
                    aWrites[i] = LJM.CONSTANTS.WRITE;
                    aNumValues[i] = 1;
                    aValues[i] = 0; //output-low
                    i++;
                }

                if(writeDACs)
                {
                    //Add analog output writes (DAC0-1)
                    for(int j = 0; j < 2; j++, i++)
                    {
                        aNames[i] = "DAC" + j;
                        aWrites[i] = LJM.CONSTANTS.WRITE;
                        aNumValues[i] = 1;
                        aValues[i] = 0.0; //0.0 V
                    }
                }

                Console.WriteLine("\nTest frames:");

                string wrStr = "";
                for(i = 0; i < numFrames; i++)
                {
                    if(aWrites[i] == LJM.CONSTANTS.READ)
                        wrStr = "READ";
                    else
                        wrStr = "WRITE";
                    Console.WriteLine("    " + wrStr + " " + aNames[i]);
                }
                Console.WriteLine("\nBeginning " + numIterations + " iterations...");


                //Initialize time variables
                double maxMS = 0;
                double minMS = 0;
                double totalMS = 0;
                double curMS = 0;
                Stopwatch sw;
                long freq = Stopwatch.Frequency;

                //eNames operations loop
                for(i = 0; i < numIterations; i++)
                {
                    sw = Stopwatch.StartNew();
                    LJM.eNames(handle, numFrames, aNames, aWrites, aNumValues, aValues, ref errAddr);
                    sw.Stop();
                    
                    curMS = sw.ElapsedTicks/(double)freq * 1000;
                    if(minMS == 0)
                        minMS = curMS;
                    minMS = Math.Min(curMS, minMS);
                    maxMS = Math.Max(curMS, maxMS);
                    totalMS += curMS;
                }

                Console.WriteLine("\n" + numIterations + " iterations performed:");
                Console.WriteLine("    Time taken: " +
                    Convert.ToDouble(totalMS).ToString("F3") + " ms");
                Console.WriteLine("    Average time per iteration: " +
                    Convert.ToDouble(totalMS/numIterations).ToString("F3") +
                    " ms");
                Console.WriteLine("    Min / Max time for one iteration: " +
                    Convert.ToDouble(minMS).ToString("F3") + " ms / " +
                    Convert.ToDouble(maxMS).ToString("F3") + " ms");

                Console.WriteLine("\nLast eNames results: ");
                for(i = 0; i < numFrames; i++)
                {
                    if(aWrites[i] == LJM.CONSTANTS.READ)
                        wrStr = "READ";
                    else
                        wrStr = "WRITE";
                    Console.WriteLine("    " + aNames[i] +" " + wrStr + " value : " + aValues[i]);
                }
            }
            catch (LJM.LJMException e)
            {
                showErrorMessage(e);
            }

            LJM.CloseAll(); //Close all handles

            Console.WriteLine("\nDone.\nPress the enter key to exit.");
            Console.ReadLine(); // Pause for user
        }
    }
}

File: VB/Testing/CRSpeedTest/CRSpeedTest.vb
'------------------------------------------------------------------------------
' CRSpeedTest.vb
'
' Performs LabJack operations in a loop and reports the timing statistics for
' the operations.
'
' support@labjack.com
'------------------------------------------------------------------------------
Option Explicit On

Imports LabJack

Module CRSpeedTest

    Sub showErrorMessage(ByVal e As LJM.LJMException)
        Console.WriteLine("LJMException: " & e.ToString)
        Console.WriteLine(e.StackTrace)
    End Sub

    Sub displayHandleInfo(ByVal handle As Integer)
        Dim devType As Integer
        Dim conType As Integer
        Dim serNum As Integer
        Dim ipAddr As Integer
        Dim port As Integer
        Dim maxBytesPerMB As Integer
        Dim ipAddrStr As String = ""

        LJM.GetHandleInfo(handle, devType, conType, serNum, ipAddr, port, _
                          maxBytesPerMB)
        LJM.NumberToIP(ipAddr, ipAddrStr)
        Console.WriteLine("Opened a LabJack with Device type: " & devType & _
                          ", Connection type: " & conType & ",")
        Console.WriteLine("Serial number: " & serNum & ", IP address: " & _
                          ipAddrStr & ", Port: " & port & ",")
        Console.WriteLine("Max bytes per MB: " & maxBytesPerMB)
    End Sub

    Sub Main()
        Dim handle As Integer
        Dim numFrames As Integer
        Dim aNames() As String
        Dim aWrites() As Integer
        Dim aNumValues() As Integer
        Dim aValues() As Double
        Dim errAddr As Integer = -1

        Dim i As Integer
        Dim wrStr As String

        Const numIterations As Integer = 1000 ' Number of iterations to perform in the loop

        Const numAIN As Integer = 1 ' Number of analog inputs to read
        Const rangeAIN As Double = 10.0
        Const resolutionAIN As Double = 1.0

        ' Digital settings
        Const readDigital As Boolean = False
        Const writeDigital As Boolean = False

        ' Analog output settings
        Const writeDACs As Boolean = False

        ' Time variables
        Dim maxMS As Double = 0
        Dim minMS As Double = 0
        Dim totalMS As Double = 0
        Dim curMS As Double = 0
        Dim sw As Stopwatch
        Dim freq As Long = Stopwatch.Frequency

        Try
            ' Open first found LabJack
            LJM.OpenS("ANY", "ANY", "ANY", handle)
            'LJM.Open(LJM.CONSTANTS.dtANY, LJM.CONSTANTS.ctANY, "ANY", handle)

            displayHandleInfo(handle)


            If numAIN > 0 Then
                ' Configure analog input settings
                numFrames = Math.Max(0, numAIN * 2)
                ReDim aNames(numFrames - 1)
                ReDim aValues(numFrames - 1)
                For i = 0 To numAIN - 1
                    aNames(i * 2) = "AIN" + i.ToString() + "_RANGE"
                    aValues(i * 2) = rangeAIN
                    aNames(i * 2 + 1) = "AIN" + i.ToString() + _
                                        "_RESOLUTION_INDEX"
                    aValues(i * 2 + 1) = resolutionAIN
                Next
                LJM.eWriteNames(handle, numFrames, aNames, aValues, errAddr)
            End If

            ' Initialize and configure eNames parameters for loop's eNames call
            numFrames = Math.Max(0, numAIN) + Convert.ToInt32(readDigital) + _
                        Convert.ToInt32(writeDigital) + _
                        Convert.ToInt32(writeDACs) * 2
            ReDim aNames(numFrames - 1)
            ReDim aWrites(numFrames - 1)
            ReDim aNumValues(numFrames - 1)
            ReDim aValues(numFrames - 1) ' In this case numFrames is the size of aValue

            ' Add analog input reads (AIN 0 to numAIN-1)
            For i = 0 To numAIN - 1
                aNames(i) = "AIN" + i.ToString()
                aWrites(i) = LJM.CONSTANTS.READ
                aNumValues(i) = 1
                aValues(i) = 0
            Next

            If readDigital Then
                ' Add digital read
                aNames(i) = "DIO_STATE"
                aWrites(i) = LJM.CONSTANTS.READ
                aNumValues(i) = 1
                aValues(i) = 0
                i += 1
            End If

            If writeDigital Then
                ' Add digital write
                aNames(i) = "DIO_STATE"
                aWrites(i) = LJM.CONSTANTS.WRITE
                aNumValues(i) = 1
                aValues(i) = 0 ' output-low
                i += 1
            End If

            If writeDACs Then
                ' Add analog output writes (DAC0-1)
                For j = 0 To 1
                    aNames(i) = "DAC" + j.ToString()
                    aWrites(i) = LJM.CONSTANTS.WRITE
                    aNumValues(i) = 1
                    aValues(i) = 0.0 ' 0.0 V
                    i += 1
                Next
            End If

                    Console.WriteLine("")
                    Console.WriteLine("Test frames:")

            wrStr = ""
            For i = 0 To numFrames - 1
                If aWrites(i) = LJM.CONSTANTS.READ Then
                    wrStr = "READ"
                Else
                    wrStr = "WRITE"
                End If
                Console.WriteLine("    " & wrStr & " " & aNames(i))
            Next
            Console.WriteLine("")
            Console.WriteLine("Beginning " & numIterations & " iterations...")


            ' eNames operations loop
            For i = 0 To numIterations - 1
                sw = Stopwatch.StartNew()
                LJM.eNames(handle, numFrames, aNames, aWrites, aNumValues, _
                           aValues, errAddr)
                sw.Stop()

                curMS = sw.ElapsedTicks / freq * 1000
                If minMS = 0 Then minMS = curMS
                minMS = Math.Min(curMS, minMS)
                maxMS = Math.Max(curMS, maxMS)
                totalMS += curMS
            Next

            Console.WriteLine("")
            Console.WriteLine(numIterations & " iterations performed:")
            Console.WriteLine("    Time taken: " & totalMS.ToString("F3") & _
                              " ms")
            Console.WriteLine("    Average time per iteration: " & _
                (totalMS / numIterations).ToString("F3") & " ms")
            Console.WriteLine("    Min / Max time for one iteration: " & _
                minMS.ToString("F3") & " ms / " & maxMS.ToString("F3") & _
                " ms")

            Console.WriteLine("")
            Console.WriteLine("Last eNames results: ")
            For i = 0 To numFrames - 1
                If aWrites(i) = LJM.CONSTANTS.READ Then
                    wrStr = "READ"
                Else
                    wrStr = "WRITE"
                End If
                Console.WriteLine("    " & aNames(i) & " " & wrStr & _
                                  " value : " & aValues(i).ToString("0.####"))
            Next
        Catch ljme As LJM.LJMException
            showErrorMessage(ljme)
        End Try

        LJM.CloseAll() ' Close all handles

        Console.WriteLine("")
        Console.WriteLine("Done.")
        Console.WriteLine("Press the enter key to exit.")
        Console.ReadLine() ' Pause for user
    End Sub

End Module
